cpp/src/utilities/column_utils.cpp

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2019 BlazingDB, Inc.
+ *     Copyright 2019 Eyal Rozenberg <eyalroz@blazingdb.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "column_utils.hpp"
+#include <utilities/error_utils.hpp>
+
+// This should go up into more general utility code...
+static constexpr inline bool logical_xor(bool x, bool y) { return (x and (not y)) or ((not x) and y) ; }
+
+namespace cudf {
+
+/**
+ * @brief Ensures a gdf_column is valid, i.e. that its fields are consistent
+ * with each other, and logical in themselves, in representing a proper column.
+ */
+void validate(const gdf_column& column)
+{
+    CUDF_EXPECTS ( (column.data != nullptr or column.size == 0)          , "No column data despite there supposedly being rows.");
+    CUDF_EXPECTS ( (column.dtype != GDF_invalid)                         , "Cannot use the 'invalid' column element type.");
+    CUDF_EXPECTS ( (column.dtype < N_GDF_TYPES)                          , "Unknown column data type");
+    CUDF_EXPECTS ( (is_nullable(column) or (column.null_count == 0))     , "Column supposedly has null values, but no place to expose its record");
+    CUDF_EXPECTS ( (column.null_count <= column.size)                    , "Column claims to have more null elements than elements overall");
+}
+
+void validate(const gdf_column* column_ptr)
+{
+    CUDF_EXPECTS ( (column_ptr != nullptr) , "Null gdf_column pointer");
+    validate(*column_ptr);
+}
+
+namespace detail {
+
+bool extra_type_info_is_compatible(
+    const gdf_dtype& common_dtype,
+    const gdf_dtype_extra_info& lhs_extra_type_info,
+    const gdf_dtype_extra_info& rhs_extra_type_info) noexcept
+{
+    switch(common_dtype) {
+    case GDF_CATEGORY:  return lhs_extra_type_info.category == rhs_extra_type_info.category;
```

Comment:
This check is problematic, and results from the hackiness in how we support the NVCategory column type.

`gdf_dtype_extra_info::category` is only set for columns of type `GDF_STRING_CATEGORY`. Furthermore, I don't think it is correct to check if two columns of type `GDF_STRING_CATEGORY` simply point to the same `NVCategory` object. Two category dictionaries can be compatible without pointing to the exact same NVCategory objects.

Presently, I do not believe there is any way to check if two `NVCategory` objects are compatible. Instead, in features like join, we explicitly synchronize the two NVCategory objects to guarantee they are compatible.

Reply:
What would you suggest, then? I have no preference one way or another here.

Reply:
I would skip comparing `lhs_extra_type_info.category` for any column type.

Reply:
@jrhemstad : Ok, done.